Introduction:
In the rapidly evolving field of data science, understanding the factors that influence job salaries is crucial for professionals and employers alike. With Python being a popular programming language in the data science community, it offers a powerful toolset for conducting comprehensive analyses. This study focuses on analyzing data science job salaries in 2023 using Python. By examining various factors such as job titles, experience levels, educational qualifications, and geographical locations, we aim to uncover insights into the salary landscape of the data science industry. The findings of this analysis will provide valuable information for both aspiring data scientists and organizations seeking to attract and retain top talent.



Data analysis has emerged as a prominent occupation in the technology industry, contributing significantly to the global economy. The exponential growth of data generation in recent years has led to a demand for professionals skilled in data science. Data science utilizes statistics, mathematics, scientific methods, computer programming, machine learning, deep learning, and other techniques to extract knowledge, uncover insights, and identify patterns from large, unstructured datasets. The expansion of this field has created numerous job opportunities worldwide, making data science one of the most sought-after career paths today. The growth trajectory of this field is expected to continue due to the increasing need for analyzing vast amounts of data. Within data science, there are diverse job titles that exist. To gain insights into these job titles and their corresponding salaries, a dataset comprising 3755 job titles from companies across the globe will be analyzed through a series of question-based analyses. This analysis aims to identify the key factors influencing the relationship between job titles and their associated salaries.

Importing Required Libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')
import plotly.express as px
import plotly.graph_objects as go

Loading the Data

In [ ]:
df=pd.read_csv('datafield_salaries.csv')
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [ ]:
# Shape of the dataset i.e. No of rows and columns
df.shape


(3755, 11)

In [ ]:
# General detailed information of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3755 entries, 0 to 3754
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           3755 non-null   int64 
 1   experience_level    3755 non-null   object
 2   employment_type     3755 non-null   object
 3   job_title           3755 non-null   object
 4   salary              3755 non-null   int64 
 5   salary_currency     3755 non-null   object
 6   salary_in_usd       3755 non-null   int64 
 7   employee_residence  3755 non-null   object
 8   remote_ratio        3755 non-null   int64 
 9   company_location    3755 non-null   object
 10  company_size        3755 non-null   object
dtypes: int64(4), object(7)
memory usage: 322.8+ KB


In [ ]:
# Statistical Information,description of the Numeric Attributes
df.describe()

work_year        salary  salary_in_usd  remote_ratio
count  3755.000000  3.755000e+03    3755.000000   3755.000000
mean   2022.373635  1.906956e+05  137570.389880     46.271638
std       0.691448  6.716765e+05   63055.625278     48.589050
min    2020.000000  6.000000e+03    5132.000000      0.000000
25%    2022.000000  1.000000e+05   95000.000000      0.000000
50%    2022.000000  1.380000e+05  135000.000000      0.000000
75%    2023.000000  1.800000e+05  175000.000000    100.000000
max    2023.000000  3.040000e+07  450000.000000    100.000000

In [ ]:
# Statistical Information of the Categorical Attributes
df.describe(include=['O'])

experience_level employment_type      job_title salary_currency  \
count              3755            3755           3755            3755   
unique                4               4             93              20   
top                  SE              FT  Data Engineer             USD   
freq               2516            3718           1040            3224   

       employee_residence company_location company_size  
count                3755             3755         3755  
unique                 78               72            3  
top                    US               US            M  
freq                 3004             3040         3153

In [ ]:
# Checking for the missing values
df.isna().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

**Data Manipulation and Feature Engineering**


Converting Country Codes (ISO3)
As part of our analysis of data science job salaries in 2023, we may need to convert country codes to the ISO3 format. The ISO3 format is a standardized three-letter country code system used to represent countries globally. This conversion allows for consistency and ease of analysis when dealing with country-related data.

To convert country codes to ISO3 format, we can utilize various methods. One common approach is to use a library or dataset that maps country codes to their corresponding ISO3 codes. This mapping can be done by matching the country codes from our dataset with the ISO3 codes provided in the library or dataset.

By performing this conversion, we can ensure uniformity in the representation of country

In [ ]:
! pip install country_converter

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
# Converting country codes of employee_residence column
import country_converter as coco
df['employee_residence'] = coco.convert(names=df['employee_residence'])
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [ ]:
# Converting country codes of company_location column
import country_converter as coco
df['company_location'] = coco.convert(names=df['company_location'])
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size  
0              ESP            L  
1              USA            S  
2              USA            S  
3              CAN            M  
4              CAN            M

Summary:

In the analysis of data science job data, a new column is added to distinguish **FOREIGN employees from NON-FOREIGN** employees based on the comparison between the employee's home country and the country where the company is based.

 If the two countries are the same, the employee is classified as a non-foreign employee, while if the countries differ, the employee is categorized as a foreign employee.

This categorization allows for a deeper understanding of the composition of foreign employees within the dataset. It provides insights into the distribution and representation of foreign workers in the data science job market/industry, which will further explored in relation to job titles, salaries, or other variables of interest.

By differentiating between foreign and non-foreign employees, this analysis enables the identification of patterns and trends associated with the employment of foreign workers in the data science industry. Such insights can be valuable for companies seeking to understand the diversity of their workforce and for professionals interested in exploring career opportunities in different countries.

In [ ]:
df['foreign_employee']=np.where(df['employee_residence']==df['company_location'],"No","Yes")
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size foreign_employee  
0              ESP            L               No  
1              USA            S               No  
2              USA            S               No  
3              CAN            M               No  
4              CAN            M               No

**Determining Currency Conversion Rates**

In the provided dataset, salaries are listed in various currencies, alongside a column indicating the corresponding salaries in USD. By leveraging this information, we can calculate the conversion rates between different currencies, using USD(stable currency) as the standard currency.

In [ ]:
df['conversion_rates'] = df['salary']/df['salary_in_usd']
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size foreign_employee  conversion_rates  
0              ESP            L               No           0.93189  
1              USA            S               No           1.00000  
2              USA            S               No           1.00000  
3              CAN            M               No           1.00000  
4              CAN            M               No           1.00000

**Elaborating Experience Level Using the Replace Method**

The dataset includes experience levels represented in a short format. To provide a more detailed representation of experience levels, will utilize the replace method to expand the abbreviated format into descriptive terms. This will enhance the clarity and comprehensibility of the experience level information within the dataset.

In [ ]:
df['experience_level'].replace(['SE', 'MI', 'EN', 'EX'],
                               ['Senior-level','Mid-level','Entry-level','Executive-level'],
                               inplace=True)
df.head()

work_year experience_level employment_type                 job_title  \
0       2023     Senior-level              FT  Principal Data Scientist   
1       2023        Mid-level              CT               ML Engineer   
2       2023        Mid-level              CT               ML Engineer   
3       2023     Senior-level              FT            Data Scientist   
4       2023     Senior-level              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size foreign_employee  conversion_rates  
0              ESP            L               No           0.93189  
1              USA            S               No           1.00000  
2              USA            S               No           1.00000  
3              CAN            M               No           1.00000  
4              CAN            M               No           1.00000

Elaborating Employment Type

To provide a more comprehensive understanding of the employment types within the dataset, will expand the abbreviated format of the employment type. By using appropriate methods, we can replace the abbreviated format with descriptive terms, enabling clearer interpretation and analysis of the employment type information.

In [ ]:
df['employment_type'].replace(['FT', 'CT', 'FL', 'PT'],['Full-time','Contract','Freelancer','Part-time'], inplace=True)
df.head()

work_year experience_level employment_type                 job_title  \
0       2023     Senior-level       Full-time  Principal Data Scientist   
1       2023        Mid-level        Contract               ML Engineer   
2       2023        Mid-level        Contract               ML Engineer   
3       2023     Senior-level       Full-time            Data Scientist   
4       2023     Senior-level       Full-time            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                ESP           100   
1   30000             USD          30000                USA           100   
2   25500             USD          25500                USA           100   
3  175000             USD         175000                CAN           100   
4  120000             USD         120000                CAN           100   

  company_location company_size foreign_employee  conversion_rates  
0              ESP            L               No           0.93189  
1              USA            S               No           1.00000  
2              USA            S               No           1.00000  
3              CAN            M               No           1.00000  
4              CAN            M               No           1.00000

# **Elaborating Job Type Using Remote Ratio**

1.   List item
2.   List item



To enhance the clarity of job types in the dataset, we can utilize the remote_ratio column to replace the values with more descriptive terms. By mapping the remote_ratio values to specific job types, we can assign labels such as "remote (100%)", "on-site (0%)", and "hybrid (50%)" to provide a clearer representation of the nature of each job. This will facilitate a better understanding of the job types and their remote work arrangements within the dataset.

In [ ]:
df['remote_ratio'].replace([0,50,100],['On-Site','Hybrid','Remote'], inplace=True)
df.rename(columns={'remote_ratio':'job_type'}, inplace=True) #Renaming the column
df.head()

work_year experience_level employment_type                 job_title  \
0       2023     Senior-level       Full-time  Principal Data Scientist   
1       2023        Mid-level        Contract               ML Engineer   
2       2023        Mid-level        Contract               ML Engineer   
3       2023     Senior-level       Full-time            Data Scientist   
4       2023     Senior-level       Full-time            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence job_type  \
0   80000             EUR          85847                ESP   Remote   
1   30000             USD          30000                USA   Remote   
2   25500             USD          25500                USA   Remote   
3  175000             USD         175000                CAN   Remote   
4  120000             USD         120000                CAN   Remote   

  company_location company_size foreign_employee  conversion_rates  
0              ESP            L               No           0.93189  
1              USA            S               No           1.00000  
2              USA            S               No           1.00000  
3              CAN            M               No           1.00000  
4              CAN            M               No           1.00000

# Question Based Analysis

**1) Which all countries took part in this survey?**

To determine the countries that participated in this survey, will need to examine the dataset and extract the unique values from the column representing the country of residence or any other relevant column. By identifying the distinct countries mentioned in the dataset, we can ascertain the countries that took part in this survey.

In [ ]:
fig = px.choropleth(locations = df['employee_residence'],
                    color = df['employee_residence'],
                    title = 'Countries involved in the Survey')
fig.show()

**2) Which are the top 10 popular job designations?**

To identify the top 10 popular job designations in the dataset, will analyze the column containing the job titles or any relevant column related to job designations. By performing a frequency count of the job titles and selecting the top 10 most frequently occurring designations, we can determine the popular job titles in the dataset.


In [ ]:
df['job_title'].value_counts().head(10)

Data Engineer                1040
Data Scientist                840
Data Analyst                  612
Machine Learning Engineer     289
Analytics Engineer            103
Data Architect                101
Research Scientist             82
Data Science Manager           58
Applied Scientist              58
Research Engineer              37
Name: job_title, dtype: int64

In [ ]:
top_10_jobs = df['job_title'].value_counts().head(10)  # Calculating the top 10 most frequent job titles

fig = px.bar(x=top_10_jobs.index,  # Setting the x-axis values as the job designations
             y=top_10_jobs,  # Setting the y-axis values as the number of posts for each job designation
             title='Top 10 Most Popular Job Designations',  # Setting the title of the chart
             labels={'y': 'No. of Posts', 'x': 'Job Designations'})  # Setting the labels for the axes

fig.show()  # Displaying the bar chart


The job designation of Data Engineer stands out as one of the most popular, closely followed by Data Scientist and Data Analyst.

# 3) Which are the top 10 highest paid job designations over the years?
To identify the top 10 highest paid job designations over the years, we can analyze the salary data in the dataset and consider the job titles associated with the highest salaries. By sorting the salaries in descending order and selecting the top 10 job designations, we can determine the job titles that have consistently commanded high salaries throughout the years.

In [ ]:
xdf=df.groupby(['job_title'])['salary_in_usd'].median().sort_values(ascending=False).head(10)
xdf


job_title
Data Science Tech Lead                 375000.0
Cloud Data Architect                   250000.0
Data Lead                              212500.0
Data Analytics Lead                    211254.5
Head of Data                           202500.0
Principal Data Engineer                192500.0
Applied Scientist                      191737.5
Principal Machine Learning Engineer    190000.0
Data Science Manager                   183780.0
Data Infrastructure Engineer           183655.0
Name: salary_in_usd, dtype: float64

In [ ]:
px.bar(x=xdf.index, y=xdf, title='Top 10 High Paying Job Designations',
      labels={'y':'Median Salary','x':'Job Designations'})

# 4) Compare the company sizes and salaries offered by them over the years.

To identify the top 10 highest paid job designations over the years, we can analyze the salary data in the dataset and consider the job titles associated with the highest salaries. By sorting the salaries in descending order and selecting the top 10 job designations, we can determine the job titles that have consistently commanded high salaries throughout the years.

In [ ]:
table = pd.pivot_table(df,values='salary_in_usd',index='work_year',columns='company_size', aggfunc='median')
table

company_size         L         M        S
work_year                                
2020           85000.0   96500.0  51321.0
2021           90734.0   64411.5  76833.0
2022          120000.0  135000.0  58418.5
2023          123700.0  145000.0  74875.5

In [ ]:
px.box(df, x='company_size', y='salary_in_usd', color='work_year',
       category_orders={"company_size":["S","M","L"],"work_year":[2020, 2021, 2022, 2023]},
      title='Company Size vs Salary')


Medium-sized companies consistently offer the highest median salaries over the years, with noticeable wide distribution in the year 2022 and 2023. On the other hand, large-sized companies exhibit a trend of increasing median salaries over time. In contrast, small-sized companies demonstrate fluctuation in their salary levels.

# 5) Which companies are preferred by employees if we consider their sizes?

If we take into account the sizes of companies, we can determine which companies are preferred by employees. By analyzing the dataset, we can examine the company sizes and identify the companies that are most sought after by employees. This analysis will provide insights into the preferences of employees based on company size and help identify the companies that are considered desirable by job seekers.

In [ ]:
company_size_counts = df['company_size'].value_counts(normalize=True) * 100

fig = go.Figure(data=[go.Pie(labels=company_size_counts.index,
                             values=company_size_counts,
                             hole=0.6,
                             title='Percentage Distribution of Company Sizes')])

fig.show()


The higher median salary could be a reason behind 84% of employees' preference to work in medium-sized companies.

# 6. What is the percentage of remote workers who are classified as foreign?

To provide the data or variables necessary to determine the number of employees working remotely, on-site, and in a hybrid capacity for each year?

In [ ]:
table = df.pivot_table(index='work_year',values='salary',columns='job_type',aggfunc='count')
table

job_type   Hybrid  On-Site  Remote
work_year                         
2020           21       16      39
2021           76       34     120
2022           62      711     891
2023           30     1162     593

In [ ]:
table=df.pivot_table(values='salary_in_usd', index='work_year', columns='job_type')
table

job_type         Hybrid        On-Site         Remote
work_year                                            
2020       77591.238095   93426.187500   99763.205128
2021       76104.407895   92899.676471  105812.783333
2022       84560.870968  134719.236287  135631.104377
2023       72053.500000  152393.521515  146380.124789

In [ ]:
px.histogram(x=df['work_year'],color=df['job_type'],barmode='group',labels={'x':'Work Year'},title='Count of Job Types wih Year')

Over the years, there has been a significant growth in the number of onsite jobs, with a notable exponential increase from 2020 to 2023. On the other hand, remote jobs experienced an upward trend from 2020 to 2022, followed by a decline in 2023. In contrast, the number of hybrid jobs remained relatively stable throughout the period, showing minimal variation.


# 7 Which job type demonstrates higher salary levels over the course of multiple years?

In [ ]:
table=df.pivot_table
table = df.pivot_table(values='salary_in_usd', index='work_year', columns='job_type')
table


job_type         Hybrid        On-Site         Remote
work_year                                            
2020       77591.238095   93426.187500   99763.205128
2021       76104.407895   92899.676471  105812.783333
2022       84560.870968  134719.236287  135631.104377
2023       72053.500000  152393.521515  146380.124789

In [ ]:
px.box(df,x='work_year', y='salary_in_usd', color='job_type',
      title='Distribution of Salaries over the Job Types with Year')

In the year 2020, as a result of the pandemic, a significant number of individuals were working remotely or in a hybrid mode. Consequently, the median salary for both remote and hybrid job types was higher compared to on-site positions. However, the median salary for on-site workers gradually increased over time and nearly equaled that of remote workers in 2022. By 2023, the median salary for on-site workers slightly surpassed that of remote workers.
It is worth noting that both on-site and remote job types exhibit a wide distribution of salaries, indicating substantial variation. In contrast, individuals working in a hybrid mode experienced less variation in their salaries.

# What is the percentage of remote workers who are classified as foreign?

In [ ]:
df['foreign_employee'].value_counts()/len(df['foreign_employee'])


No     0.974434
Yes    0.025566
Name: foreign_employee, dtype: float64

In [ ]:
px.pie(values=df['foreign_employee'].value_counts(),
       names=df['foreign_employee'].value_counts().index,
      title='Proportion of Foreign Employees')


Out of the total individuals surveyed, only 2.56% are classified as foreign employees.

# 9 In which country do the majority of foreign employees work?

In [ ]:
xdf=df[df['foreign_employee']=='Yes']['company_location'].value_counts()
xdf.head(10)

USA    41
DEU    10
GBR     6
AUS     5
CAN     4
SGP     2
LUX     2
FRA     2
ASM     2
ESP     2
Name: company_location, dtype: int64

In [ ]:
fig = px.choropleth(locations = xdf.index, color = xdf, title = 'Countries with Remote Foreign Employees')
fig.show()

Most foreing employees work in USA followed by Germany.

10) Which country has the most number of people working for foreign companies?

In [ ]:
xdf=df[df['foreign_employee']=='Yes']['employee_residence'].value_counts()
xdf.head(10)


IND    15
FRA     6
USA     5
PAK     5
ESP     5
ITA     4
PRT     4
BRA     3
ARG     3
POL     2
Name: employee_residence, dtype: int64

In [ ]:
import plotly.graph_objects as go
#table = df.pivot_table(index='work_year', values='salary', columns='job_type', aggfunc='count')

fig = go.Figure()

for job_type in table.columns:
    fig.add_trace(go.Bar(
        x=table.index,
        y=table[job_type],
        name=job_type
    ))

fig.update_layout(
    title='Comparison of Job Types by Work Year',
    xaxis_title='Work Year',
    yaxis_title='Count of Job Types'
)

fig.show()


In [ ]:
fig = px.pie(values=df['foreign_employee'].value_counts(),
             names=df['foreign_employee'].value_counts().index,
             title='Proportion of Foreign Employees',
             hole=0.5)

fig.show()


In [ ]:
xdf=df[df['foreign_employee']=='Yes']['company_location'].value_counts()
xdf.head(10)

USA    41
DEU    10
GBR     6
AUS     5
CAN     4
SGP     2
LUX     2
FRA     2
ASM     2
ESP     2
Name: company_location, dtype: int64

In [ ]:
fig = px.choropleth(locations = xdf.index, color = xdf, title = 'Countries with People working in Foreign Country')
fig.show()

India tops the list of No. of people working for a foreign country followed by France.


# 11 To determine the weakest and strongest currencies, we can compare their conversion rates to USD as the standard currency. By analyzing the variations in currency exchange rates, we can identify which currencies are relatively stronger or weaker compared to the US dollar.

In [ ]:
#Top 5 Strongest Currencies
curr_stgh=df.groupby(['salary_currency'])['conversion_rates'].median()

In [ ]:
curr_stgh.sort_values().head(5)

salary_currency
GBP    0.812134
EUR    0.931899
CHF    0.955135
USD    1.000000
CAD    1.301833
Name: conversion_rates, dtype: float64

In [ ]:
xdf=df[df.conversion_rates<1.36]
px.box(xdf, x='salary_currency',y='conversion_rates')

12) What is the number of peole working in the field of data science in each country?

Since we have taken USD as the standard currency it doesn't show any variation. On the other hand EUR and CAD show much variation. And GBP continues to be storngest currency throughout.

In [ ]:
# Top 5 Weakest Currencies
curr_stgh.sort_values(ascending=False).head(5)

salary_currency
CLP    759.278685
HUF    307.821464
JPY    109.870224
INR     78.627701
THB     33.952756
Name: conversion_rates, dtype: float64

In [ ]:
xdf=df[df.conversion_rates<1.36]
px.box(xdf, x='salary_currency',y='conversion_rates')

CLP stands for Chilean Peso, which remains the weakest currency with minimal fluctuations. On the other hand, HUF represents the Hungarian Forint, which exhibits significant variations in its strength.


12. What is the count of individuals working in the field of data science in each country?

In [ ]:
xdf = df['employee_residence'].value_counts().sort_values(ascending=False)
xdf.head(10)

USA    3004
GBR     167
CAN      85
ESP      80
IND      71
DEU      48
FRA      38
PRT      18
BRA      18
GRC      16
Name: employee_residence, dtype: int64

In [ ]:
fig = px.choropleth(locations = xdf.index, color = xdf, title = 'No. of People working in the Field of Data Science')
fig.show()

The United States has the highest number of individuals working in the field of data science.

13. I would like to compare the salaries based on the levels of experience and years of work.

In [ ]:
table = df.pivot_table(values='salary_in_usd',columns='work_year',index='experience_level')
table

work_year                  2020           2021           2022           2023
experience_level                                                            
Entry-level        57511.608696   54905.254545   77006.024194   95283.966102
Executive-level   139944.333333  186128.000000  188260.292683  203705.683333
Mid-level          87564.718750   82116.934783  101305.598338  116297.596875
Senior-level      137240.500000  126085.356164  147659.688049  159568.928516

In [ ]:
px.box(df,x='experience_level',y='salary_in_usd', color='work_year',
       category_orders={"work_year":[2020, 2021, 2022, 2023],
                       "experience_level":["Entry-level","Mid-level","Senior-level", "Executive-level"]},
       title='Experience Level vs Salary')

Over the years, there is a consistent increase in the median salary across all levels of experience. Furthermore, senior-level and executive-level professionals tend to receive higher median salaries compared to individuals at other experience levels.

14. I would like to compare the relationship between employment type and salary.


In [ ]:
xdf = df.groupby('employment_type', as_index=False)['salary_in_usd'].median()
xdf

employment_type  salary_in_usd
0        Contract        75000.0
1      Freelancer        50000.0
2       Full-time       135000.0
3       Part-time        21669.0

In [ ]:
px.bar(xdf,x='employment_type',y='salary_in_usd')

Full-time employees earn higher salaries compared to any other employment type.

15. What is the median salary offered to employees from each country?

In [ ]:
xdf = df.groupby('employee_residence')['salary_in_usd'].median().sort_values(ascending=False)
xdf.head()

employee_residence
ISR    423834.0
MYS    200000.0
PRI    160000.0
USA    145000.0
CHN    125404.0
Name: salary_in_usd, dtype: float64

In [ ]:
fig = px.choropleth(locations = xdf.index, color = xdf, title = 'Salary offered by Each Country to an Employee')
fig.show()

What is the median salary offered by each company based on their location?

In [ ]:
xdf = df.groupby('company_location')['salary_in_usd'].median().sort_values(ascending=False)
xdf.head()

company_location
ISR    271446.5
PRI    167500.0
USA    145000.0
NZL    125000.0
BIH    120000.0
Name: salary_in_usd, dtype: float64

In [ ]:
fig = px.choropleth(locations = xdf.index, color = xdf, title = 'Salary offered by Each Company')
fig.show()

17. I would like to visualize the distribution of salaries across different years.

In [ ]:
px.histogram(x=df['salary_in_usd'], color=df['work_year'], barmode='overlay',
            labels={'x':'Salary'}, title='Distribution of Salaries over the Years')

The distribution of salaries is right-skewed, meaning that there are fewer values with higher magnitudes.

18. I would like to analyze the number of job posts over the years.

In [ ]:
xdf=df['work_year'].value_counts()
xdf


2023    1785
2022    1664
2021     230
2020      76
Name: work_year, dtype: int64

In [ ]:
px.pie(values=xdf, names=xdf.index, title='Proportion of No. of Posts')

# The year 2023 has the highest number of job posts, followed by the year 2022. This indicates significant growth in the field, and it is expected to continue expanding. If you found this notebook helpful, please consider upvoting it. For any suggestions or inquiries, please feel free to reach out. Let's progress together. Thank you.

 *This concludes our discussion. If you have any further questions, feel free to ask.*

Since we have taken USD as the standard currency it doesn't show any variation. On the other hand EUR and CAD show much variation. And GBP continues to be storngest currency throughout.

There is always an increment over the years in median salary for all the experience levels. Seniors and Executives get higher median salary than the others.

# 14) Compare the employment type with salary.

In [ ]:
xdf = df.groupby('employment_type', as_index=False)['salary_in_usd'].median()
xdf

employment_type  salary_in_usd
0        Contract        75000.0
1      Freelancer        50000.0
2       Full-time       135000.0
3       Part-time        21669.0

In [ ]:
px.bar(xdf,x='employment_type',y='salary_in_usd')

# 15) What is the median salary offered to an employee from each country?

In [ ]:
xdf = df.groupby('employee_residence')['salary_in_usd'].median().sort_values(ascending=False)
xdf.head()

employee_residence
ISR    423834.0
MYS    200000.0
PRI    160000.0
USA    145000.0
CHN    125404.0
Name: salary_in_usd, dtype: float64

In [ ]:
fig = px.choropleth(locations = xdf.index, color = xdf, title = 'Salary offered by Each Country to an Employee')
fig.show()

16) What is the median salary offered by each company depending on the company location?

In [ ]:
xdf = df.groupby('company_location')['salary_in_usd'].median().sort_values(ascending=False)
xdf.head()

company_location
ISR    271446.5
PRI    167500.0
USA    145000.0
NZL    125000.0
BIH    120000.0
Name: salary_in_usd, dtype: float64

In [ ]:
fig = px.choropleth(locations = xdf.index, color = xdf, title = 'Salary offered by Each Company')
fig.show()

17) Show the distribution of salaries over the years

In [ ]:
px.histogram(x=df['salary_in_usd'], color=df['work_year'], barmode='overlay',
            labels={'x':'Salary'}, title='Distribution of Salaries over the Years')

The distribution of salaries is rightly skewed i.e. fewer number of values have higher magnitudes.

# 18) Analyze the number of job posts over the years.

In [ ]:
xdf=df['work_year'].value_counts()
xdf

2023    1785
2022    1664
2021     230
2020      76
Name: work_year, dtype: int64

In [ ]:
px.pie(values=xdf, names=xdf.index, title='Proportion of No. of Posts')

The year 2023 contributes maximum number of posts followed by year 2022. We can see that the field is growing immensely and will continue to do.

If you find this notebook useful please upvote and if you have any sugesstions or queries please feel free to contact me. Let's grow together. Thank you.

The End¶